frontend/Obelisk/OAuth/Frontend.hs

```diff
@@ -0,0 +1,217 @@
+    --   Nothing when not yet authorized, else error or actual token.
+  }
+
+makeReflexLenses ''OAuthFrontend
+
+
+type OAuthConstraints t m =
+  ( Reflex t, MonadIO m, MonadHold t m, MonadJSM m, PostBuild t m
+  , PerformEvent t m , MonadJSM (Performable m)
+  )
+
+data StoreOAuth a where
+  -- State that gets stored to session storage:
+  StoreOAuth_State :: StoreOAuth OAuthState
+  -- Access token that gets stored to local storage:
+  StoreOAuth_Token :: StoreOAuth OAuthToken
+
+deriving instance Show (StoreOAuth a)
+
+-- | Function for asking the backend for the actual access token.
+type TokenGetter m = RedirectUriParams -> m (Either OAuthError OAuthToken)
+
+
+-- | Make an `OAuthFrontend` by reading `OAuthConfig` by means of `getOAuthConfigPublic`.
+--
+--   TODO: We probably want to get rid of this function/ change type signature
+--   a bit. As we'd like to encourage the user to provide a config generation
+--   function based on `getOAuthConfigPublic` with user values already
+--   provided, which can then be used both in frontend and backend. We also
+--   might want to consider moving more values into the config. (Ideally we
+--   would just not need some of those all together.)
+makeOAuthFrontendExeCfg
+  :: OAuthConstraints t m
+  => OAuthProvider
+  -> Encoder Identity Identity (R (Sum r a)) PageName
+  -> Maybe (r (R OAuthRoute))
+  -> TokenGetter (Performable m)
+  -> OAuthFrontendConfig t
+  -> m (OAuthFrontend t)
+makeOAuthFrontendExeCfg provider enc redirectUri getToken cfg = do
+  sCfg <- getOAuthConfigPublic provider redirectUri
+  makeOAuthFrontend enc sCfg getToken cfg
+
+
+-- | Make an `OAuthFrontend` provided with the needed configuration.
+makeOAuthFrontend
+  :: OAuthConstraints t m
+  => Encoder Identity Identity (R (Sum r a)) PageName
+  -> OAuthConfigPublic r
+  -> TokenGetter (Performable m)
+  -> OAuthFrontendConfig t
+  -> m (OAuthFrontend t)
+makeOAuthFrontend enc sCfg getToken cfg = do
+  onErrParams <- retrieveCode enc sCfg (_oAuthFrontendConfig_route cfg) $ _oAuthFrontendConfig_authorize cfg
+  eToken <- retrieveToken sCfg getToken onErrParams
+  pure $ OAuthFrontend eToken
+
+
+-- | Takes care of issuing the initial authorization request and handling the
+--   code response.
+--
+--   State will be properly generated and checked.
+retrieveCode
+  :: forall t m r a. OAuthConstraints t m
+  => Encoder Identity Identity (R (Sum r a)) PageName
+  -> OAuthConfigPublic r
+  -> Dynamic t (Maybe (R OAuthRoute))
+  -> Event t ()
+  -> m (Event t (Either OAuthError RedirectUriParams))
+retrieveCode enc sCfg route onAuth = do
+    performEvent_ $ doAuthorize enc sCfg <$ onAuth
+
+    onRoute <- tagOnPostBuild route
+    mOldState <- getItemStorage sessionStorage StoreOAuth_State
+    let
+      onParamsErr = fmapMaybe (fmap (checkState mOldState =<<) . getParams) onRoute
+    performEvent_ $ removeItemStorage sessionStorage StoreOAuth_State <$ onParamsErr
```

Comment:
I don't think it makes sense for this package to worry about how the oauth-related data is persisted.  Let's let the caller specify whatever's needed to store this, e.g. in local storage or somewhere else.

Reply:
Also, can you explain what's going on here with removing the storage item and such?

Reply:
Well, as we redirect the user to github - we need to store the generated state somewhere, so we have it when the user comes back. I picked session storage as it is ideal for this purpose.

I remove it from storage mostly for good measure to ensure any kind of replay attacks are not possible. It should not be necessary as session storage is transient anyway.

Letting the user specify what's needed to store this is most likely overkill, also customization could be added later on when the need arises in a backwards compatible way. (Assuming users don't do a full pattern match on the config of course.

Reply:
I believe the risk of the user using the same key in session storage for something else is pretty low, although we should probably add to the documentation what keys are used in session storage by default. - Will do!